src/hugr/views/petgraph.rs

```diff
@@ -1,187 +1,250 @@
+{
+    type Neighbors = <T as HugrView>::Neighbours<'g>;
+
+    fn neighbors(self, n: Self::NodeId) -> Self::Neighbors {
+        self.output_neighbours(n)
+    }
+}
+
+impl<'g, 'a, T> pv::IntoNeighborsDirected for &'g PetgraphWrapper<'a, T>
+where
+    T: HugrView,
+{
+    type NeighborsDirected = <T as HugrView>::Neighbours<'g>;
+
+    fn neighbors_directed(
+        self,
+        n: Self::NodeId,
+        d: petgraph::Direction,
+    ) -> Self::NeighborsDirected {
+        self.neighbours(n, d.into())
+    }
+}
+
+impl<'a, T> pv::Visitable for PetgraphWrapper<'a, T>
+where
+    T: HugrView,
+{
+    type Map = std::collections::HashSet<Self::NodeId>;
+
+    fn visit_map(&self) -> Self::Map {
+        std::collections::HashSet::new()
+    }
+
+    fn reset_map(&self, map: &mut Self::Map) {
+        map.clear();
+    }
+}
+
+impl<'a, T> pv::GetAdjacencyMatrix for PetgraphWrapper<'a, T>
+where
+    T: HugrView,
+{
+    type AdjMatrix = std::collections::HashSet<(Self::NodeId, Self::NodeId)>;
 
-            fn reset_map(&self, map: &mut Self::Map) {
-                map.clear();
+    fn adjacency_matrix(&self) -> Self::AdjMatrix {
+        let mut matrix = std::collections::HashSet::new();
+        for node in self.nodes() {
+            for neighbour in self.output_neighbours(node) {
+                matrix.insert((node, neighbour));
             }
         }
+        matrix
+    }
 
-        impl<'a, Root, Base> pv::GetAdjacencyMatrix for $hugr<'a, Root, Base>
-        where
-            Root: NodeHandle,
-            Base: HugrInternals + HugrView,
-        {
-            type AdjMatrix = std::collections::HashSet<(Self::NodeId, Self::NodeId)>;
-
-            fn adjacency_matrix(&self) -> Self::AdjMatrix {
-                let mut matrix = std::collections::HashSet::new();
-                for node in self.nodes() {
-                    for neighbour in self.output_neighbours(node) {
-                        matrix.insert((node, neighbour));
-                    }
-                }
-                matrix
-            }
+    fn is_adjacent(&self, matrix: &Self::AdjMatrix, a: Self::NodeId, b: Self::NodeId) -> bool {
+        matrix.contains(&(a, b))
+    }
+}
 
-            fn is_adjacent(
-                &self,
-                matrix: &Self::AdjMatrix,
-                a: Self::NodeId,
-                b: Self::NodeId,
-            ) -> bool {
-                matrix.contains(&(a, b))
-            }
+impl<'a, T> HugrInternals for PetgraphWrapper<'a, T>
+where
+    T: HugrView,
+{
+    type Portgraph<'p> = <T as HugrInternals>::Portgraph<'p>
+    where
+        Self: 'p;
+
+    delegate! {
+        to self.hugr {
+            fn portgraph(&self) -> Self::Portgraph<'_>;
+            fn base_hugr(&self) -> &crate::Hugr;
+            fn root_node(&self) -> Node;
         }
-    };
+    }
 }
 
-impl_region_petgraph_traits!(SiblingGraph);
-impl_region_petgraph_traits!(DescendantsGraph);
+impl<'g, T> HugrView for PetgraphWrapper<'g, T>
```

Comment:
Do we really need this? If you've got the PetgraphWrapper then you've got the view, no? (TBH I'd expect to be given the HugrView to start off with, then construct the PetgraphWrapper as/when needed, i.e. only when passing off to petgraph methods that don't care about HugrView, but there)

Reply:
Note that if PetgraphWrapper implements HugrView then there is the possibility of nested PetgraphWrapper's, which I think we should probably avoid!?

Reply:
Yeah, this isn't necessary. I'll drop it.